plugins/woocommerce/src/Internal/Utilities/BatchProcessingController.php

```diff
@@ -0,0 +1,287 @@
+		$pending_updates = $this->get_pending();
+		if ( ! in_array( $processor_class_name, array_keys( $pending_updates ) ) ) {
+			$pending_updates[] = $processor_class_name;
+			$this->set_pending_processes( $pending_updates );
+		}
+		$this->schedule_init_cron();
+	}
+
+	/**
+	 * Helper method to start update cron.
+	 *
+	 * @param bool $with_delay Whether to delay the cron start. Send true when rescheduling, false when starting.
+	 */
+	private function schedule_init_cron( bool $with_delay = false ) {
+		$time = $with_delay ? time() + HOUR_IN_SECONDS : time();
+		as_schedule_single_action(
+			$time,
+			self::CONTROLLER_CRON_NAME,
+			$this->get_args_for_init_cron(),
+			self::ACTION_GROUP
+		);
+	}
+
+	/**
+	 * Schedules update for all updaters that may be stuck. This method is called in CONTROLLER_CRON_NAME action.
+	 */
+	public function schedule_processes() {
+		$pending_processes = $this->get_pending();
+		if ( empty( $pending_processes ) ) {
+			return;
+		}
+		foreach ( $pending_processes as $process_name ) {
+			if ( ! $this->already_scheduled( $process_name ) ) {
+				$this->schedule_next_batch( $process_name );
+			}
+		}
+		$this->schedule_init_cron( true );
+	}
+
+	/**
+	 * Process update for a scheduled updater.
+	 *
+	 * @param string $batch_process Fully qualified class name of the updater. Must be child class `BatchProcessor`.
+	 */
+	public function process_single_batch( string $batch_process ) {
+		$batch_processor = $this->get_processor_instance( $batch_process );
+		$this->process_batch( $batch_processor );
+		if ( $batch_processor->get_total_pending_count() > 0 ) {
+			$this->schedule_next_batch( $batch_process );
+		} else {
+			$this->mark_pending_process_complete( $batch_process );
+			$batch_processor->mark_process_complete();
+		}
+	}
+
+	/**
+	 * Process next batch for given instance of `BatchProcessor`.
+	 *
+	 * @param BatchProcessor $batch_processor Batch processor instance.
+	 *
+	 * @return bool True if batch was processed, false if not.
+	 */
+	private function process_batch( BatchProcessor $batch_processor ) {
+		$details    = $this->get_process_details( $batch_processor );
+		$time_start = microtime( true );
+		$batch      = $batch_processor->get_batch_data( $details['current_batch_size'], $details['last_processed'] );
+		if ( empty( $batch ) ) {
+			return false;
+		}
+		try {
+			$batch_processor->process_for_batch( $batch );
+			$time_taken = microtime( true ) - $time_start;
+			$this->update_progress_status( $batch_processor, $batch, $time_taken );
+			return true;
+		} catch ( \Exception $exception ) {
+			$this->log_error( $exception );
+			$this->update_progress_status( $batch_processor, $batch, $time_taken, $exception );
+			return false;
+		}
+	}
+
+	/**
+	 * Get status for this update.
+	 *
+	 * @param BatchProcessor $batch_processor Batch processor instance.
+	 *
+	 * @return mixed Update status.
+	 */
+	protected function get_process_details( BatchProcessor $batch_processor ) {
```

Comment:
Do this method and the companion `update_progress_status` really need to be protected? I can't think of any use case for "messing" with the batch enqueueing/scheduling process in a derived class, and if what you want is to provide visibility for the sake of logging/data visualization, I think that a couple of hooks would do a better job here.

Reply:
Agreed, modified to make then `private` in https://github.com/woocommerce/woocommerce/pull/33233/files#diff-77d9531091cce5f175d5bb8fe3a1560d9dda936d370c81fabb90ca0f46633e52R170